KIPs/kip-146.md

```diff
@@ -0,0 +1,171 @@
+
+## Motivation
+
+Klaytn has adopted the Byzantine Fault Tolerance (BFT) consensus mechanism, where a block proposer selected by a deterministic algorithm proposes the next block. A vulnerable point of BFT is that an adversary can effectively halt the entire network by simply targeting the proposer, a.k.a. targeted DoS. What makes Klaytn even more vulnerable to a targeted DoS attack is that the current proposer selection policy enables the prediction of up to `proposerUpdateInterval` (3600 in case of Cypress) number of proposers in advance.
+
+As of now, a targeted DoS is practically infeasible in Klaytn because it is a permissioned chain which allows only authorized validators to join the network. However, in a permissionless network, a targeted DoS is a viable option for attackers. A random proposer selection can increase the difficulty of the attack by introducing an uncertainty of the upcoming proposers.
+
+Meanwhile, a few malicious validators must not be able to forge the randomness to arbitrarily determine the upcoming proposers. If not, it may increase the risk of censorship or monopolizing block rewards. Therefore, the new random proposer selection policy should be both unpredictable and verifiable.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Shuffling
+
+A shuffling algorithm must be defined which will act as a building block of proposer and committee selection.
+
+[Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) must be used in the below pseudocode (i.e., `random.shuffle`), and [Golang's RNG](https://cs.opensource.google/go/go/+/master:src/math/rand/rng.go;l=7-12) must be used as a pseudo random number generator internally in Fisher-Yates.
+
+The shuffling function must be deterministic given `validators` and `seed`.
+
+```py
+def shuffle_validators_KIP146(validators: list[int], seed: int):
+    """returns a list of randomly shuffled validators.
+    Keyword arguments:
+    validators -- deterministically sorted list of validators.
+    seed -- `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` of the previous block is used (see KIP-114).
+    """
+    ret = validators[:]
+
+    random.seed(seed)
+    random.shuffle(ret)
+    return ret
+```
+
+### Proposer selection
+
+For calculating the proposer of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
```

Comment:
Can we change WeightedRandom to UnpredictableRandom on the fork block?

Reply:
Let's mention it in docs, instead of here